app/controllers/twitter-searcher.controller.js

```diff
@@ -0,0 +1,130 @@
+(function() {
+  'use strict';
+
+  class TwitterSearcherCtrl {
+    constructor($q, $resource, BarGraphService, MapGraphService, LinearGraphService, Oboe) {
+      Object.assign(this, {$q, $resource, BarGraphService, MapGraphService, LinearGraphService, Oboe});
+      this.tweeters = [];
+      this.streams = [];
+      this.words = [
+      { id: 1,
+        word: ''
+      }];
+    }
+
+    add(word, inputNumber) {
+
+      this.streams = [];
+      if (this.words.length < 4) {
+        this.words.push({
+          id: this.words.length + 1,
+          word: ''
+        });
+      }
+
+      var currentWords = [];
+      this.words.forEach((element) => {
+        currentWords.push(element.word);
+      });
+
+      var params = {
+        action: 'search',
+        query: word
+      }
+
+      this.tweets = this.$resource('/tweets/:action/:query', params);
+
+      this.tweets.query({}, (res) => {
+        if(this.tweeters[inputNumber - 1]) {
+          return this.tweeters.splice(inputNumber - 1, 1, res);
+        }
+        this.tweeters.push(res);
+      });
+
+      oboe('/tweets/filter/' + currentWords.join())
+        .node('$!.created_at', (tweet) => {
+          this.streams.push(tweet);
+          this.groupStreamByWord(this.streams, currentWords);
+        });
+    }
+
+    groupStreamByWord(stream, words) {

```

Comment:
w tej metodzie dzieje sie zdecydowanie za duzo rzeczy:
. czytanie ze streama,
. parsowanie danych,
. tworzenie struktur danych,
. rysowanie wykresow.


wszystkie te rzeczy rozbij do osobnych funkcji (czasem wydzielajac je poza kontroler) a pozniej polacz w jedna metode wywolujac w niej te funkcje i trzymajac glowny flow. to sposob na realizacje Single Responsibility z SOLID